### DISCUSSION TITLE
Identifying IBAN as one token

### DISCUSSION THREAD

**Discussion by JBOE22175** [explosion/spaCy#8858] [Help: Coding & Implementations] [answered]
I try to identify the IBAN (International Bank Account Number) as one token and set the entity to IBAN. 
The following works but I wonder if there is an easier way to do it:
```
import spacy
import re

nlp = spacy.load("de_core_news_sm")
doc = nlp("Die IBAN ist DE55 1234 0000 1234 1234 99")

iban_pattern = r"DE\d{2}\ ?\d{4}\ ?\d{4}\ ?\d{4}\ ?\d{4}\ ?\d{2}"
for match in re.finditer(iban_pattern, doc.text):
    start, end = match.span()
    span = doc.char_span(start, end)
    # This is a Span object or None if match doesn't map to valid token sequence
    if span is not None:
        print("Found match:", span.text)
        print(span.start, span.end, span.ents)
```

```
Found match: DE55 1234 0000 1234 1234 99
3 9 [DE55]
```

`print("Before",[token.text for token in doc])`
`Before ['Die', 'IBAN', 'ist', 'DE55', '1234', '0000', '1234', '1234', '99']`

```
with doc.retokenize() as retokenizer:
    retokenizer.merge(span, attrs={"ENT_ID": "IBAN"})
print("After:", [token.text for token in doc])
```

`After: ['Die', 'IBAN', 'ist', 'DE55 1234 0000 1234 1234 99']`

```
for token in doc:
    print(f'{token.text:35} {token.lemma_:35}  {token.ent_id_:10}')
```

```
Die                                 der                                            
IBAN                                IBAN                                           
ist                                 sein                                           
DE55 1234 0000 1234 1234 99         DE55 1234 0000 1234 1234 99          IBAN  
```

**Comment by DavidGerva** [comment]
I don't get exactly what you want to accomplish.
If you are just trying to build a map of entity - position, your code can do the job.

I just have two main doubts:

1. do you want to train a model to find IBAN in a text? I would raise some concerns about this... can't you just use your regex? 
2. if you want to train a model with spacy >= 3.0 you have to create the corpus with the "new" method: https://towardsdatascience.com/using-spacy-3-0-to-build-a-custom-ner-model-c9256bea098

**Reply by JBOE22175** [reply]
Thank you.
I do e-mail classification and just want to replace the IBAN by some fixed text, e.g. "IBAN". Regex seems to be appropriate. See also my reply to polm.

**Comment by polm** [comment, marked as answer]
I think your approach looks fine. There was a similar thread (#8854) about phone numbers recently. If you need to deal with tokens that have unpredictable tokenization but are easy to match with regexes I think that using `char_span` like this is the way to go.

**Reply by JBOE22175** [reply]
Ok, thank you. 
Fortunately, IBAN has is a fixed regual expression. 
I used doc.retokenize() to combine the IBAN into one token. As I just want to classify the document (=e-mail) the IBAN only needs to be replaced by a the fixed word "IBAN". This probably indicates that the sender wants to inform me about the change of his/her banking account. 
Now, I tried to do this in a script before (!) spacy tokenization with
`text = re.sub(r"DE\d{2}\ ?\d{4}\ ?\d{4}\ ?\d{4}\ ?\d{4}\ ?\d{2}", 'IBAN',text)`
This works fine.
Do you have an idea which approach is better, e.g. in terms of runtime, spacy templates?

By the way: 
Do you have a good reference for text classification (e.g. e-mails) with some focus on preprocessing. The spacy documentation starts with cleaned texts but in real life the e-mails are "dirty" and must be cleaned before using as input.

**Reply by polm** [reply]
In general preprocessing texts for spaCy is a bad idea. In the old days before running a text classifier or something making everything lower case and removing stop words was standard practice, but with modern methods that's not necessary and in some cases can break things. 

Your example here is one of the cases where preprocessing can still help, though it may not help as much as you think. Internally spaCy uses "word shape" features, so it knows things like "huh, this document has a bunch of four-digit numbers in a row" that are useful for downstream tasks. 

> Do you have an idea which approach is better, e.g. in terms of runtime, spacy templates?

Not entirely sure what you're asking but I think it's fine to run this as a plain regex before feeding your text to spaCy. That would also be faster than creating a doc, then using the regex, then creating another doc. 

If the user ever uses the word "IBAN", you might want to make the token special by using `xxxIBANxxx` or something.

> Do you have a good reference for text classification (e.g. e-mails) with some focus on preprocessing. The spacy documentation starts with cleaned texts but in real life the e-mails are "dirty" and must be cleaned before using as input.

In my experience every application or data source has their own particular, special set of problems and there's no such thing as generic preprocessing. What I would recommend is just use your raw, messy text to get some baseline performance numbers and very carefully measure changes in performance as you try introducing preprocessing steps. Use some small models so you can iterate very quickly on this - you can always bring out the bigger models later.

Another thing is to look at data augmentation techniques; spaCy uses some basic case normalization so changes in case aren't that important, but there are lot of other techniques you can use to make your models more robust (= less weird with dirty data).

**Reply by JBOE22175** [reply]
Thankx again for your helpful comments.

Most posts/tutorials (even from 2019/2020) about text/e-mail classification write about the importance of preprocessing/cleaning of the text. Perhaps they are already outdated.

**My baseline** was "spacy.textcat_multilabel_demo" (which uses spacy.TextCatBOW.v1) with just 52 unclean e-Mails (70% for training and 30% for test) and three classes. 

- The result w/o preprocessing was poor (50-70% precision and recall). 
- After preprocessing I achieve 75-100% precision and 50-100% recall. 

**Preprocessing steps** include substitution of banking data (IBAN, BIC), e-Mail adresses and Links. Lower all tokens, remove exeeding whitespaces, punctuation, new lines and tabs and remove stopwords. Remove greetings at beginning and end of mail as well as text corpus of previous mails in case it is an answer to a previous mail. 

My idea was that a bag of words model will not benefit from any of the removed tokes. 
I guess that your advice applies to transformer models (but perhaps not to bag of words models). 
Do you agree?

**generic preprocessing:**
I have experience in some areas of ML but it is my first project with NLP. 
Indeed I dreamed of a template for e-mail classification (should not be too exotic) _from raw extracted e-mails_ to _multi-label-classification_ as a starting point. Changing some preprocessing steps and setting the class lables for the trainging data could give one a good baseline with little effort.

**Reply by polm** [reply]
> Most posts/tutorials (even from 2019/2020) about text/e-mail classification write about the importance of preprocessing/cleaning of the text. Perhaps they are already outdated.

They aren't outdated, but a lot of people write using old best practices without questioning them. For example if you want to use Naive Bayes, which is traditionally very popular for spam filters, it will still work and will benefit from stopword removal, but if you're using more recent techniques you don't have to do that.

> My idea was that a bag of words model will not benefit from any of the removed tokes.
I guess that your advice applies to transformer models (but perhaps not to bag of words models).
Do you agree?

That's correct in general. Bag of words models don't benefit from stop words, and removing stopwords will break Transformers. However even without Transformers models can depend on word order, for example the default spaCy models have CNNs that get some contextual information. 

Also note that with word vectors, starting with word2vec, the vector generation process almost always includes techniques that automatically handle stop words based on frequency, so there's no need to treat them specially. They'll get vectors with low magnitude and just kind of wash out in any final classification. 

Going over the list of things you did in preprocessing, I think that removing previous mails was probably very important, and lowercasing and stopword removal were probably less important. 

If you have a method that works, though, there's no need to change it.

> Indeed I dreamed of a template for e-mail classification (should not be too exotic) from raw extracted e-mails to multi-label-classification as a starting point. Changing some preprocessing steps and setting the class lables for the trainging data could give one a good baseline with little effort.

That could be a good example project. The [20 newsgroups dataset](http://qwone.com/~jason/20Newsgroups/) is a well-known classification dataset based on email, though I think most people work from a standard cleaned-up copy. The categories are unlikely to be similar to anything someone would want in practice though (the labels are the original newsgroups).

**Reply by JBOE22175** [reply]
Very helpful answer. I will try several approches. Thank you very much!
